doc/DOCUMENTATION.org

```diff
@@ -542,6 +543,58 @@ configuration at the beginning and end of =Spacemacs= loading process.
     loading.
   - =dotspacemacs/config= is triggered at the very end of =Spacemacs= loading.
 
+*** Binding keys
+Key sequences are bound to commands in Emacs in various keymaps. The most basic
+map is the global-map. Setting a key binding the global-map uses the function
+=global-set-key= as follows (to the command =forward-char= in this case).
+
+#+begin_src emacs-lisp
+  (global-set-key (kbd "C-]") 'forward-char)
+#+end_src
+
+The =kbd= macro accepts a string describing a key sequence. The global-map is
+often shadowed by other maps. For example, evil-mode defines keymaps that target
+states (or modes in vim terminology). Here is an example that creates the same
+binding as above but only in insert state (=define-key= is a built-in function.
+Evil-mode has its own functions for defining keys).
+
+#+begin_src emacs-lisp
+  (define-key evil-insert-state-map (kbd "C-]") 'forward-char)
+#+end_src
+
+Perhaps most importantly for spacemacs is the use of the evil-leader package,
+which binds keys to the evil-leader keymap. This is where most of the spacemacs
+bindings live. There are two related commands from this package which are used
+as follows.
+
+#+begin_src emacs-lisp
+  (evil-leader/set-key "C-]" 'forward-char)
+  (evil-leader/set-key-for-mode 'emacs-lisp-mode "C-]" 'forward-char)
+#+end_src
+
+These functions use a macro like =kbd= to translate the key sequences for you.
+The second function, =evil-leader/set-key-for-mode=, binds the key only in the
+specified mode. The second key binding would not be in effect in =org-mode= for
+example.
+
+Finally, one should be aware of prefix keys. Essentially, all keymaps can be
+nested. Nested keymaps are used extensively in spacemacs, and in vanilla Emacs
+for that matter. For example, ~SPC a~ points to key bindings for "applications",
+like ~SPC ac~ for =calc-dispatch=. Nesting bindings is easy.
+
+#+begin_src emacs-lisp
+  (spacemacs/declare-prefix "]" "bracket-prefix")
+  (evil-leader/set-key "]]" 'double-bracket-command)
+#+end_src
```

Comment:
Hmm. Still doesn't feel quite right. It says "Creating prefix keys is easy.". I'd say it's so easy that you're always just doing that implicitly, and never explicitly. You're not going "I'm going to create a prefix key." and write some code. You just create a key binding for more than one key. The focus is still the function the key sequence is bound to. The prefix just works automatically.

Maybe the name `spacemacs/declare-prefix` makes it sound a little "I first need to declare this thing, then I can use it." – like a variable in certain programming languages. But in fact declare-prefix just sets a which-key name for the prefix that helps documentation. The prefix itself works automatically, and you don't have to declare it.

"nesting keymaps" or "nesting bindings" sounds like this to me:

```elisp
(keymap/prefix "]" "brackety things"
  ("]" 'double-bracket-command
   ")" 'bracket-paren-command
   "-" 'bracket-dash-command))
```

But that's not what it is. Or am I missing something about prefixes?

Maybe something like this (if you use it, feel free to change at will up to the point where it's unrecognizable :)):

"You can also bind commands to sequences of keys. This is used extensively in spacemacs to group related key bindings together. You can define a name for any key sequence with `spacemacs/declare-prefix`. This name will be used in the popup window that shows possible continuations of the current key sequence."

I actually just looked at `spacemacs/declare-prefix`, and I think it does a lot more than setting a name for the prefix to be used by `which-key`. I'm new to emacs and elisp. It calls `define-prefix-command`. Hmm, does a key sequence binding work only if all prefixes in the binding have been defined with `define-prefix-command`? Maybe the problem here is that I don't really understand the thing that is being documented :)